File: sdk/python/cerberus_ai/classifiers/mcp_scanner.py
"""
cerberus_ai.classifiers.mcp_scanner
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
MCP Tool-Description Poisoning scanner — L2 sub-classifier.

Scans MCP (or generic) tool descriptions for hidden instructions,
cross-tool manipulation, data-routing commands, and zero-width
obfuscation. Based on Invariant Labs' tool-poisoning research.

Modes:
    ``scan_tool_descriptions(tools)`` — standalone registration-time scan.
    ``check_tool_call_poisoning(tool_name, tools)`` — runtime per-call lookup.

Parity port of ``src/classifiers/mcp-scanner.ts``. Same pattern set,
same category → severity mapping, same return shape.
"""
from __future__ import annotations

import json
import re
from collections.abc import Iterable
from dataclasses import dataclass

from cerberus_ai.models import ToolSchema

_POISONING_PATTERNS: tuple[tuple[str, re.Pattern[str]], ...] = (
    # Hidden instruction tags
    ("hidden_instruction", re.compile(r"<IMPORTANT>", re.I)),
    ("hidden_instruction", re.compile(r"<SYSTEM>", re.I)),
    ("hidden_instruction", re.compile(r"<ADMIN>", re.I)),
    ("hidden_instruction", re.compile(r"<!--[\s\S]*?-->", re.M)),

    # Sensitive file references
    ("sensitive_file_ref", re.compile(r"~\/\.ssh\b")),
    ("sensitive_file_ref", re.compile(r"~\/\.env\b")),
    ("sensitive_file_ref", re.compile(r"~\/\.aws\b")),
    ("sensitive_file_ref", re.compile(r"\/etc\/passwd\b")),
    ("sensitive_file_ref", re.compile(r"~\/\.gnupg\b")),
    (
        "sensitive_file_ref",
        re.compile(r"(?:credentials|secrets)\.(?:json|yaml|yml|toml|env)", re.I),
    ),

    # Cross-tool manipulation
    ("cross_tool_manipulation", re.compile(r"before\s+calling\b", re.I)),
    (
        "cross_tool_manipulation",
        re.compile(r"instead\s+of\s+(?:using|calling)\b", re.I),
    ),
    ("cross_tool_manipulation", re.compile(r"after\s+using\b", re.I)),
    ("cross_tool_manipulation", re.compile(r"when\s+.*\btool\b", re.I)),

    # Instruction injection
    ("instruction_injection", re.compile(r"\byou\s+must\b", re.I)),
    (
        "instruction_injection",
        re.compile(
            r"\balways\b.*\b(?:send|forward|include|attach)\b", re.I
        ),
    ),
    (
        "instruction_injection",
        re.compile(r"\bnever\b.*\b(?:tell|inform|reveal|show)\b", re.I),
    ),
    (
        "instruction_injection",
        re.compile(r"\bignore\b.*\b(?:instructions|rules|policies)\b", re.I),
    ),
    ("instruction_injection", re.compile(r"\boverride\b", re.I)),

    # Data routing
    ("data_routing", re.compile(r"\bsend\s+to\b", re.I)),
    ("data_routing", re.compile(r"\bforward\s+to\b", re.I)),
    ("data_routing", re.compile(r"\bcopy\s+to\b", re.I)),
    ("data_routing", re.compile(r"\bexfiltrate\b", re.I)),
    ("data_routing", re.compile(r"\bupload\s+(?:to|all)\b", re.I)),

    # Zero-width obfuscation
    ("obfuscation", re.compile(r"[\u200B\u200C\uFEFF\u00AD]")),
)

_HIGH_RISK = {"hidden_instruction", "data_routing", "obfuscation"}
_MEDIUM_RISK = {"sensitive_file_ref", "cross_tool_manipulation"}


@dataclass(frozen=True)
class ToolPoisoningResult:
    tool_name: str
    poisoned: bool
    patterns_found: tuple[str, ...]
    severity: str  # "low" | "medium" | "high"


def scan_description(text: str) -> list[str]:
    """Return the de-duplicated list of matched poisoning categories."""
    matched: set[str] = set()
    for category, pattern in _POISONING_PATTERNS:
        if pattern.search(text):
            matched.add(category)
    return sorted(matched)


def _determine_severity(patterns: Iterable[str]) -> str:
    patterns = list(patterns)
    if not patterns:
        return "low"
    if any(p in _HIGH_RISK for p in patterns):
        return "high"
    if any(p in _MEDIUM_RISK for p in patterns):
        return "medium"
    return "low"


def scan_tool_descriptions(
    tools: Iterable[ToolSchema],
) -> list[ToolPoisoningResult]:
    """Standalone scan — call at registration time for every declared tool."""
    results: list[ToolPoisoningResult] = []
    for tool in tools:
        description = tool.description or ""
        patterns = set(scan_description(description))
        if tool.parameters:
            patterns.update(scan_description(json.dumps(tool.parameters)))
        patterns_found = sorted(patterns)
        results.append(
            ToolPoisoningResult(
                tool_name=tool.name,
                poisoned=len(patterns_found) > 0,
                patterns_found=tuple(patterns_found),
                severity=_determine_severity(patterns_found),
            )
        )
    return results


def check_tool_call_poisoning(
    tool_name: str, tools: Iterable[ToolSchema]
) -> ToolPoisoningResult | None:
    """Runtime scan — return the poisoning verdict for the tool being called.

    ``None`` if the tool name isn't in the declared set (silent pass)."""
    for tool in tools:
        if tool.name != tool_name:
            continue
        description = tool.description or ""
        patterns = set(scan_description(description))
        if tool.parameters:
            patterns.update(scan_description(json.dumps(tool.parameters)))
        patterns_found = sorted(patterns)
        return ToolPoisoningResult(
            tool_name=tool_name,
            poisoned=len(patterns_found) > 0,
            patterns_found=tuple(patterns_found),
            severity=_determine_severity(patterns_found),
        )
    return None

File: src/classifiers/mcp-scanner.ts
/**
 * MCP Tool Description Poisoning Scanner — Sub-classifier enhancing L2.
 *
 * Scans MCP tool descriptions for hidden instructions, sensitive file
 * references, cross-tool manipulation, and obfuscation. Based on Invariant
 * Labs research (84.2% tool poisoning attack success rate).
 *
 * Two modes:
 * 1. Standalone: scanToolDescriptions() — called at registration time
 * 2. Runtime: checkToolCallPoisoning() — called per-tool-call if configured
 */

import type {
  ToolDescription,
  ToolPoisoningResult,
  ToolPoisoningSignal,
} from '../types/signals.js';
import type { ToolCallContext } from '../types/context.js';
import type { DetectionSession } from '../engine/session.js';

/** Poisoning pattern with category. */
interface PoisoningPattern {
  readonly category: string;
  readonly pattern: RegExp;
}

/** Patterns for detecting tool description poisoning. */
const POISONING_PATTERNS: readonly PoisoningPattern[] = [
  // Hidden instruction tags
  { category: 'hidden_instruction', pattern: /<IMPORTANT>/gi },
  { category: 'hidden_instruction', pattern: /<SYSTEM>/gi },
  { category: 'hidden_instruction', pattern: /<ADMIN>/gi },
  { category: 'hidden_instruction', pattern: /<!--[\s\S]*?-->/g },

  // Sensitive file references
  { category: 'sensitive_file_ref', pattern: /~\/\.ssh\b/g },
  { category: 'sensitive_file_ref', pattern: /~\/\.env\b/g },
  { category: 'sensitive_file_ref', pattern: /~\/\.aws\b/g },
  { category: 'sensitive_file_ref', pattern: /\/etc\/passwd\b/g },
  { category: 'sensitive_file_ref', pattern: /~\/\.gnupg\b/g },
  {
    category: 'sensitive_file_ref',
    pattern: /(?:credentials|secrets)\.(?:json|yaml|yml|toml|env)/gi,
  },

  // Cross-tool manipulation
  { category: 'cross_tool_manipulation', pattern: /before\s+calling\b/gi },
  { category: 'cross_tool_manipulation', pattern: /instead\s+of\s+(?:using|calling)\b/gi },
  { category: 'cross_tool_manipulation', pattern: /after\s+using\b/gi },
  { category: 'cross_tool_manipulation', pattern: /when\s+.*\btool\b/gi },

  // Instruction injection in descriptions
  { category: 'instruction_injection', pattern: /\byou\s+must\b/gi },
  {
    category: 'instruction_injection',
    pattern: /\balways\b.*\b(?:send|forward|include|attach)\b/gi,
  },
  { category: 'instruction_injection', pattern: /\bnever\b.*\b(?:tell|inform|reveal|show)\b/gi },
  {
    category: 'instruction_injection',
    pattern: /\bignore\b.*\b(?:instructions|rules|policies)\b/gi,
  },
  { category: 'instruction_injection', pattern: /\boverride\b/gi },

  // Data routing commands
  { category: 'data_routing', pattern: /\bsend\s+to\b/gi },
  { category: 'data_routing', pattern: /\bforward\s+to\b/gi },
  { category: 'data_routing', pattern: /\bcopy\s+to\b/gi },
  { category: 'data_routing', pattern: /\bexfiltrate\b/gi },
  { category: 'data_routing', pattern: /\bupload\s+(?:to|all)\b/gi },

  // Zero-width character obfuscation
  { category: 'obfuscation', pattern: /\u200B|\u200C|\uFEFF|\u00AD/g },
];

/**
 * Scan a single tool description for poisoning patterns.
 * Returns matched categories.
 */
export function scanDescription(text: string): string[] {
  const matched = new Set<string>();

  for (const { category, pattern } of POISONING_PATTERNS) {
    pattern.lastIndex = 0;
    if (pattern.test(text)) {
      matched.add(category);
    }
  }

  return [...matched];
}

/**
 * Determine severity based on pattern categories found.
 */
function determineSeverity(patterns: readonly string[]): 'low' | 'medium' | 'high' {
  const highRisk = ['hidden_instruction', 'data_routing', 'obfuscation'];
  const mediumRisk = ['sensitive_file_ref', 'cross_tool_manipulation'];

  if (patterns.some((p) => highRisk.includes(p))) {
    return 'high';
  }
  if (patterns.some((p) => mediumRisk.includes(p))) {
    return 'medium';
  }
  return 'low';
}

/**
 * Standalone scan — check all tool descriptions for poisoning at registration time.
 */
export function scanToolDescriptions(
  tools: readonly ToolDescription[],
): readonly ToolPoisoningResult[] {
  const results: ToolPoisoningResult[] = [];

  for (const tool of tools) {
    const patternsFound = scanDescription(tool.description);

    // Also scan parameter descriptions if present
    if (tool.parameters) {
      const paramJson = JSON.stringify(tool.parameters);
      const paramPatterns = scanDescription(paramJson);
      for (const p of paramPatterns) {
        if (!patternsFound.includes(p)) {
          patternsFound.push(p);
        }
      }
    }

    results.push({
      toolName: tool.name,
      poisoned: patternsFound.length > 0,
      patternsFound,
      severity: patternsFound.length > 0 ? determineSeverity(patternsFound) : 'low',
    });
  }

  return results;
}

/**
 * Runtime scan — check current tool call against configured tool descriptions.
 * Emits L2 signal if the tool being called has a poisoned description.
 */
export function checkToolCallPoisoning(
  ctx: ToolCallContext,
  toolDescriptions: readonly ToolDescription[],
  _session: DetectionSession,
): ToolPoisoningSignal | null {
  const toolDesc = toolDescriptions.find((t) => t.name === ctx.toolName);
  if (!toolDesc) {
    return null;
  }

  const patternsFound = scanDescription(toolDesc.description);
  if (patternsFound.length === 0) {
    return null;
  }

  return {
    layer: 'L2',
    signal: 'TOOL_POISONING_DETECTED',
    turnId: ctx.turnId,
    toolName: ctx.toolName,
    patternsFound,
    severity: determineSeverity(patternsFound),
    timestamp: ctx.timestamp,
  };
}
